### PR TITLE
[Merged by Bors] - Remove `Sync` bound from `Local`

### PR DESCRIPTION
**Pull request by PROMETHIA-27**
# Objective

Currently, `Local` has a `Sync` bound. Theoretically this is unnecessary as a local can only ever be accessed from its own system, ensuring exclusive access on one thread. This PR removes this restriction.

## Solution

- By removing the `Resource` bound from `Local` and adding the new `SyncCell` threading primative, `Local` can have the `Sync` bound removed.

### Added

- Added `SyncCell` to `bevy_utils`

### Changed

- Removed `Resource` bound from `Local`
- `Local` is now wrapped in a `SyncCell`

## Migration Guide

- Any code relying on `Local<T>` having `T: Resource` may have to be changed, but this is unlikely.

### CODE REVIEW

File: crates/bevy_utils/src/synccell.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+impl<T: ?Sized> SyncCell<T> {
+    /// Get a reference to this `SyncCell`'s inner value.
+    pub fn get(&mut self) -> &mut T {
+        &mut self.inner
+    }
+
+    /// Build a mutable reference to a `SyncCell` from a mutable reference
+    /// to its inner value, to skip constructing with [`new()`](SyncCell::new()).
+    pub fn from_mut(r: &'_ mut T) -> &'_ mut SyncCell<T> {
+        // SAFETY: repr is ≥ C, so refs have the same layout; and `SyncCell` properties are `&mut`-agnostic
```
Comment (alice-i-cecile) [inline crates/bevy_utils/src/synccell.rs:31]: I don't understand the >= C bit? The `repr` is transparent here?

Comment (PROMETHIA-27) [reply crates/bevy_utils/src/synccell.rs:31]: Me neither, actually, I just pulled that from the `Exclusive` implementation and assumed it meant something to the lang devs.

**Comment by mockersf** [comment]
bors try

File: crates/bevy_utils/src/synccell.rs (hunk 1 above)
Comment (JoJoJet) [reply crates/bevy_utils/src/synccell.rs:31]: You can just change it to say "repr is transparent", it still applies.

File: crates/bevy_utils/src/synccell.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        unsafe { &mut *(r as *mut T as *mut SyncCell<T>) }
+    }
+}
+
```
Comment (JoJoJet) [inline crates/bevy_utils/src/synccell.rs:35]:
```suggestion

// SAFETY: `Sync` only allows multithreaded access via immutable reference.
// As `SyncCell` requires an exclusive reference to access the wrapped value,
// marking this type as `Sync` does not actually allow threaded access to the inner value.
```

**Review by JoJoJet — COMMENTED** [review]
I like this a lot. Even though the use-case seems niche, it'll feel really good for that niche user when they realize it just works.

File: crates/bevy_utils/src/synccell.rs (hunk 1 above)
Comment (PROMETHIA-27) [inline crates/bevy_utils/src/synccell.rs:31]:
```suggestion
        // SAFETY: repr is transparent, so refs have the same layout; and `SyncCell` properties are `&mut`-agnostic
```

**Review by JoJoJet — APPROVED** [review]

**Comment by hymm** [comment]
I like the relaxation on the bounds for Locals, but I'm unsure about taking on `SyncCell` when `Exclusive` seems to be close-ish to being stabilized. Are there any clear use cases for this? If there are it'd probably be worth merging now.

**Comment by PROMETHIA-27** [comment]
The idea is to just swap `SyncCell` for `Exclusive` once it stabilizes. As this is probably going to be the only use of `SyncCell` until that point, it shouldn't take more than a minute to put together a PR for.

**Comment by hymm** [comment]
users could also depend on it. see the issue on `bevy_fundsp` that is in this feed.

**Comment by PROMETHIA-27** [comment]
In that case, we can typedef `SyncCell` to `Exclusive` and deprecate it, and remove it in the next major version. Or just deprecate it. Or just remove it- bevy users are still very much supposed to expect breaking changes, as long as we add to the migration guide for that version that `SyncCell` is replaced by `Exclusive` it should be fine. Either way I don't see it as worth waiting for something to stabilize "soon" unless we have a concrete date.

**Comment by harudagondi** [comment]
> see the issue on bevy_fundsp that is in this feed.

btw, i don't need it anymore as #5819 is now merged.

also, if in the case that users would need it, i would prefer just removing it outright as bevy's api isn't really stabilized yet. I think bevy's development is fast enough to update `SyncCell` usage to `Exclusive`, and we don't have to wait for stabilization.

**Review by james7132 — APPROVED** [review]
One note about a followup to this PR, otherwise LGTM.

**Pull request closed without merging**